src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java

```diff
@@ -85,8 +120,54 @@
   }
 
   private static Object merge(Object mergedStructure, Object unmergedStructure) {
-    if (unmergedStructure instanceof FieldSpec) {
-      FieldSpec fieldSpec = (FieldSpec) unmergedStructure;
+    if (unmergedStructure instanceof Spec) {
+      return ((Spec) unmergedStructure).merge(mergedStructure);
+    } else if (unmergedStructure instanceof InitValueConfig) {
+      // Only valid to merge an InitValueConfig if the existing merged structure is
+      // an empty InitValueConfig
+      if (mergedStructure instanceof InitValueConfig) {
+        InitValueConfig metadata = (InitValueConfig) mergedStructure;
+        if (metadata.isEmpty()) {
+          // Replace empty structure with unmergedStructure
+          return unmergedStructure;
+        }
+      }
+      throw new IllegalArgumentException(
+          "Inconsistent: found both substructure and initialization metadata");
+    } else {
+      throw new IllegalArgumentException(
+          "merge: didn't expect "
+              + unmergedStructure.getClass().getName()
+              + "; mergedStructure: "
+              + mergedStructure
+              + "; unmergeredStructure: "
+              + unmergedStructure);
+    }
+  }
+
+  private static Object populate(Object unmergedStructure) {
+    if (unmergedStructure instanceof Spec) {
+      return ((Spec) unmergedStructure).populate();
+    } else {
+      return unmergedStructure;
+    }
+  }
+
+  interface Spec {
+    Object merge(Object mergedStructure);
+
+    Object populate();
+  }
+
+  @AutoValue
+  abstract static class FieldSpec implements Spec {
```

Comment:
Since the Spec classes have become heavy, I think we should pull them out into separate files.